### PR TITLE
Add an fmpq_mat module

### PR DESCRIPTION
**Pull request by wbhart**

### CODE REVIEW

**Comment by wbhart** [comment]
I think the reason it shows up patches from the previous PR is because I started work on this before merging from Nemocas after my last PR was accepted.

I think the way we are supposed to avoid this is to develop our new features in private branches in our local repo, and when we merge from Nemocas, we do a git merge --rebase instead of git merge. Of course this can only be done for branches we are developing locally, which we haven't yet pushed to our GitHub repos.

File: src/flint/fmpq_mat.jl [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+
+###############################################################################
+#
+#   Inversion
+#
+###############################################################################
+
+function inv(x::fmpq_mat)
+   z = parent(x)()
+   success = ccall((:fmpq_mat_inv, :libflint), Int,
```
Comment (thofma) [inline src/flint/fmpq_mat.jl]: Return value should be `CInt`.

File: src/flint/fmpq_mat.jl [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+###############################################################################
+#
+#   Trace
+#
+###############################################################################
+
+function trace(x::fmpq_mat)
+   rows(x) != cols(x) && error("Not a square matrix in trace")
+   d = fmpq()
+   ccall((:fmpq_mat_trace, :libflint), Int,
```
Comment (thofma) [inline src/flint/fmpq_mat.jl]: Return value should be `Void`.

**Comment by wbhart** [comment]
Both correct observations, thanks. I'll change them.

**Comment by thofma** [comment]
Looks good. Just a quick (partly related) question.
```julia
julia> typeof([fmpq(2) 3 5; 1 4 7; 9 6 3])
Array{Nemo.fmpq,2}
```
I would expect the type to be `Array{Any, 2}`. Is there some promotion going on? Probably this has always been the case, but I just noticed it when skimming over tests.

**Comment by wbhart** [comment]
Yeah it does promotion of all the entries.

**Pull request merged**